### PR TITLE
Suggestions how to make CMMotionManager extensions more stateless.

### PR DESCRIPTION
**Pull request by kzaher**
Here are some ideas how we can maybe make these extensions simpler and more declarative.

### CODE REVIEW

File: Pod/Classes/RxCoreMotion.swift [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+//
+//  Created by Carlos García on 23/05/16.
+//  Copyright (c) 2016 RxSwiftCommunity  https://github.com/RxSwiftCommunity. All rights reserved.
+//
+
+import CoreMotion
+import RxSwift
+
+extension CMMotionManager {
+    static public func rx_manager(createMotionManager: () throws -> CMMotionManager = { CMMotionManager() }) -> Observable<CMMotionManager> {
```
Comment (carlosypunto) [inline Pod/Classes/RxCoreMotion.swift]: @kzaher I'm always open to your sugestions (you're my great mentor in RP). The only thing that worries me here is that Apple suggests using a single instance of CMMotionManager across the whole application.
> An app should create only a single instance of the CMMotionManager class. Multiple instances of this class can affect the rate at which data is received from the accelerometer and gyroscope.

How could we do to fix it?

**Comment by GuyKahlon** [comment]
Hey @carlosypunto and @kzaher.

I think (IMHO) there is another problem with this improvement.
 
Example:

If someone will set a again to the block "startAccelerometerUpdatesToQueue" the observer will stop working.

I've added the code below at the end of viewDidAppear on the example project (and as expected it's stopped working).

    do {
      let man = try! coreMotionManager.toBlocking().first()
      if let m = man {       
        m.startAccelerometerUpdatesToQueue(NSOperationQueue(), withHandler: { (data: CMAccelerometerData?, error: NSError?) -> Void in
          print(data)
        })
      }
    }
 
I know it's a little wired used but it can happen,  so you need some way to handle it.

I'm writing  here because I'm working on Rx wrapper for [Starscreaml](https://github.com/daltoniam/Starscreaml) called [RxStarscream](https://github.com/GuyKahlon/RxStarscream), and @kzaher refer me to this improvement branch. 

I think we have a common problem any may you can help, and maybe we can define the best practice how to create a Rx wrapper.

@kzaher Thanks again for your mentoring.

**Comment by kzaher** [comment]
Hi guys, 

I've talked with Carlos in private regarding some of his concerns, but it would make sense to make those discussions public.

I think the main thing is 
> ... Apple suggests ..

... and not

> ... Apple limits ...

I don't think it's up to us to change the generality of Apple interfaces more then we need to. By doing so, we are limiting our users.

Thnx Carlos for pointing that out, for that reasons there is a `shareReplayLatestWhileConnected` operators in the end of all wrappers.

That operator will ensure using only one CMMotionManager per Observable sequence, even if you are using multiple subsequences like `rx_acceleration`, `rx_rotationRate`, etc ...

You can also just apply any of the `retry` operators when an error occurs, and everything will work out of the box.

For the `retry` operators to work successfully, you there needs to be an appropriate reference counting mechanism in place. I understand Carlos has done that in a more direct way, but it's IMHO better to reuse existing functionalities of library that were designed for those use cases.

I didn't go all the way, and create the most perfect solution for this. There can be things that could be improved.

The intention of this PR is to demonstrate an idea how to share a resource (CMMotionManager, WebSocket ...) that needs to be turned on/off to produce elements (startDeviceMotionUpdatesToQueue, connect ....).

One would create a static method that is more or less a factory method.

If takes a configuration lambda that describes how to create a shared resource:

```swift
extension CMMotionManager {
   static public func rx_manager(createMotionManager: () throws -> CMMotionManager = { CMMotionManager() }) -> Observable<CMMotionManager>
}
```

At the end of that implementation is `.shareReplayLatestWhileConnected()` operator that controls lifetime of that shared resource. When first observer arrives, it turns it on, and when there is nobody observing, it shuts down the resource (`CMMotionManager`, `WebSocket`, .... etc ...).

To select wanted subsequence, one would then do a `flatMapLatest`, and select appropriate subsequence.

```swift
coreMotionManager
             .flatMapLatest { manager in
                 return manager.rx_rotationRate
             }
```

That's the basic idea of the approach.

Like @GuyKahlon noted, this would need to be further improved, and it can be ofc.


If `CMMotionManager` is used as an element, then some closure could turn on/off `startAccelerometerUpdatesToQueue`  by itself, and thus disrupt this reference counting mechanism, and that is bad ofc.

This is also one of the reasons why it's always preferable to have immutable sequence elements. If sequence elements are mutable, then sideeffects could be again propagated uncontrolled from any part of the system to any other. There are no isolation compile time guarantees to help us.

But we can solve this ofc :)

The simplest solution ofc is to create a dummy struct that looks something like this:

```swift
// If the current device supports one of the capabilities, observable sequence will not be nil
public struct MotionManager {
    public let acceleration: Observable<CMAcceleration>?
    public let rotationRate: Observable<CMRotationRate>?
    public let magneticField: Observable<CMMagneticField>?
    public let deviceMotion: Observable<CMDeviceMotion>?

    private init(motionManager: CMMotionManager) {
        if motionManager.accelerometerAvailable {
            self.acceleration = motionManager.rx_acceleration
        }
        else {
            self.acceleration = nil
        }

        if motionManager.gyroAvailable {
            self.rotationRate = motionManager.rx_rotationRate
        }
        else {
            self.rotationRate = nil
        }

        if motionManager.magnetometerAvailable {
            self.magneticField = motionManager.rx_magneticField
        }
        else {
            self.magneticField = nil
        }

        if motionManager.deviceMotionAvailable {
            self.deviceMotion = motionManager.rx_deviceMotion
        }
        else {
            self.deviceMotion = nil
        }
    }
}
```

And use this as a return value ...

```swift
extension CMMotionManager {
    static public func rx_manager(createMotionManager: () throws -> CMMotionManager = { CMMotionManager() }) -> Observable<MotionManager>
}
```

I hope this solves all of your concerns.

Maybe there is some more elegant solution, but so far this is the best I know how to do.

¯\\\_(ツ)\_/¯

**Comment by carlosypunto** [comment]
Forgive me that I had a week really busy

**Pull request merged**